packages/ember-application/lib/system/application.js

```diff
@@ -318,24 +364,25 @@ var Application = Namespace.extend(RegistryProxy, {
     registerLibraries();
     logLibraryVersions();
 
-    // Start off the number of deferrals at 1. This will be
-    // decremented by the Application's own `initialize` method.
+    // Start off the number of deferrals at 1. This will be decremented by
+    // the Application's own `boot` method.
     this._readinessDeferrals = 1;
 
     if (isEnabled('ember-application-visit')) {
-      if (this.autoboot) {
-        // Create subclass of Ember.Router for this Application instance.
-        // This is to ensure that someone reopening `App.Router` does not
-        // tamper with the default `Ember.Router`.
-        // 2.0TODO: Can we move this into a globals-mode-only library?
-        this.Router = (this.Router || Router).extend();
-        this.buildDefaultInstance();
-        this.waitForDOMReady();
-      }
+      this.autoboot = this._globalsMode = !!this.autoboot;
     } else {
-      this.Router = (this.Router || Router).extend();
-      this.buildDefaultInstance();
-      this.waitForDOMReady();
+      // Force-assign these flags to their default values when the feature is
+      // disabled, this ensures we can rely on their values in other paths.
+      this.autoboot = true;
+      this._globalsMode = true;
+    }
+
+    if (this._globalsMode) {
+      this._prepareForGlobalsMode();
+    }
+
+    if (this.autoboot) {
+      this._scheduleAutoBoot();
```

Comment:
If I understand the flow for boot() correctly, out of the box we are still assuming an app will wait for DOM to be ready before booting. 

If an application wants to use the public boot() function to not wait for DOM ready, it would still need to touch the constructor to call boot():
init() {
  this._super(...arguments);
  this.boot();
}

Is that correct?

My two cents:
If the above is correct, this still seems a bit hacky to me to touch the constructor to call boot(). Should we just expose a flag that determines if the application wants to wait for DOM ready or not? By default the flag is set to true.
If the flag is set to false, the constructor can call _bootSync(). 

Also we could assume the flag will only be used when the autoboot is set to true.  So probably we could update this to:
if (this.autoboot) {
  if (this.waitForDomReady) {
    // wait for DOM to be ready before booting
    this._scheduleAutoBoot();
  } else {
    // boot without waiting for DOM ready.
    this._bootSync();
  }
}